minetest/minetest
# Revamp minetest.after, option 2

This is a proposal and an alternative to https://github.com/minetest/minetest/pull/11971. The new feature of this PR is that jobs which expire simultaneously expire in the order of their registration. This is the intuitive thing to do, and it doesn't increase the complexity of the algorithm. In fact, if many jobs are inserted to expire at the same time, which sometimes happens\*, the complexity of inserting/removing each of these jobs approaches O(1).

\* E.g. when a piston extends and pushes many nodes.

## To do

This PR is Ready for Review.

## How to test

You can test it in the same way as https://github.com/minetest/minetest/pull/11971.

You can apply this patch to see the number of jobs compared to the number of items actually in the heap:

```patch
diff --git a/builtin/common/after.lua b/builtin/common/after.lua
index 94e9dbf89..26665d34c 100644
--- a/builtin/common/after.lua
+++ b/builtin/common/after.lua
@@ -100,6 +100,7 @@ end
 local time = 0.0
 local time_next = math.huge
 
+local nj = 0
 core.register_globalstep(function(dtime)
 	time = time + dtime
 
@@ -107,6 +108,8 @@ core.register_globalstep(function(dtime)
 		return
 	end
 
+minetest.chat_send_all(("Jobs: %d,  Expiries: %d"):format(nj, #expiries))
+
 	-- Remove the expired jobs.
 	local expired = remove_first_jobs()
 
@@ -122,6 +125,7 @@ core.register_globalstep(function(dtime)
 	-- Run the callbacks afterward to prevent infinite loops with core.after(0, ...).
 	local last_expired = expired.list_end
 	while true do
+nj = nj - 1
 		core.set_last_run_mod(expired.mod_origin)
 		expired.func(unpack(expired, 1, expired.n_args))
 		if expired == last_expired then
@@ -156,6 +160,7 @@ function core.after(after, func, ...)
 	local expiry = time + after
 	add_job(expiry, new_job)
 	time_next = math.min(time_next, expiry)
+nj = nj + 1
 
 	return setmetatable(new_job, job_metatable)
 end
```
